apps/io-profile-async/src/functions/expired-session-advisor.ts

```diff
@@ -0,0 +1,206 @@
+      E.mapLeft(
+        _ =>
+          new QueueTransientError(
+            "Unexpected response from function profile [retrieveProfile]"
+          )
+      )
+    ),
+    TE.chain(({ status, value }) =>
+      status === 200
+        ? TE.right(value)
+        : TE.left(
+            new QueueTransientError(
+              `Error while retrieving user profile: downstream component returned ${status} [retrieveProfile]`
+            )
+          )
+    )
+  );
+
+export const buildMailBody = (
+  expiredSessionEmailParameters: EmailParameters
+): MailBody =>
+  pipe(
+    mailTemplate.apply(
+      "TODO: replace with the real template" as NonEmptyString,
+      "TODO: replace with the real template" as NonEmptyString,
+      new Date(),
+      "TODO: replace with the real template" as NonEmptyString,
+      "TODO: replace with the real template" as NonEmptyString
+    ),
+    emailHtml => ({
+      emailHtml,
+      emailText: htmlToText(
+        emailHtml,
+        expiredSessionEmailParameters.htmlToTextOptions
+      )
+    })
+  );
+
+export const notifySessionExpiration: (
+  email: EmailAddress,
+  expiredSessionEmailParameters: EmailParameters
+) => RTE.ReaderTaskEither<
+  MailerTransporterDependency & { logger: L.Logger },
+  QueueTransientError,
+  undefined
+> = (email: EmailAddress, expiredSessionEmailParameters: EmailParameters) => ({
+  mailerTransporter,
+  logger
+}) =>
+  pipe(
+    buildMailBody(expiredSessionEmailParameters),
+    TE.of,
+    TE.chainW(({ emailHtml, emailText }) =>
+      pipe(
+        sendMail(mailerTransporter, {
+          from: expiredSessionEmailParameters.from,
+          html: emailHtml,
+          subject: expiredSessionEmailParameters.title,
+          text: emailText,
+          to: email
+        })
+      )
+    ),
+    TE.orElseW(error =>
+      pipe(
+        L.errorRTE(
+          `Error while sending email to the user => ${error.message} stack: ${error.stack}`
+        )({ logger }),
+        TE.chainW(_ =>
+          TE.left(
+            new QueueTransientError(
+              "Error while sending email to the user [notifySessionExpiration]"
+            )
+          )
+        )
+      )
+    ),
+    TE.map(_ => void 0)
+  );
+
+export const ExpiredSessionAdvisorHandler: (
+  expiredSessionEmailParameters: EmailParameters
+) => H.Handler<
+  ExpiredSessionAdvisorQueueMessage,
+  undefined,
+  BackendInternalClientDependency &
+    FunctionProfileClientDependency &
+    MailerTransporterDependency
+> = (expiredSessionEmailParameters: EmailParameters) =>
+  H.of(({ fiscalCode }: ExpiredSessionAdvisorQueueMessage) =>
+    pipe(
+      retrieveSession(fiscalCode),
+      RTE.chainW(userSessionInfo =>
+        pipe(
+          userSessionInfo.active,
+          B.fold(
+            () => RTE.right(fiscalCode), // User has no active session
+            () => RTE.left(new Error("User has an active session"))
+          )
+        )
+      ),
+      RTE.chainW(retrieveProfile),
```

Comment:
We can simplify using the `filterOrElse` method 
```suggestion
      RTE.filterOrElseW(
        userSessionInfo => !userSessionInfo.active,
        () => new Error("User has an active session")
      ),
      RTE.chainW(() => retrieveProfile(fiscalCode),
```

Reply:
new Error is wrong, should be new QueuePermanentError, anyway edited thanks for the nit pik!

Reply:
It was right at the time when I wrote the comment, obviously 😄